src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -117,11 +124,61 @@ public Http2Connection(HttpConnectionPool pool, Stream stream)
             _maxConcurrentStreams = int.MaxValue;
             _pendingWindowUpdate = 0;
 
+            if (_pool.EnableMultipleHttp2Connections)
+            {
+                _freeStreamSlots = InitialSlotsLimit;
+            }
+
             if (NetEventSource.IsEnabled) TraceConnection(stream);
         }
 
         private object SyncObject => _httpStreams;
 
+        public bool AcquireStreamSlot()
+        {
+            while (true)
+            {
+                int currentSlots = _freeStreamSlots;
```

Comment:
On architectures without x86's visibility guarantees, this could spin endlessly by not seeing a change done by another thread.

```suggestion
                int currentSlots = Volatile.Read(ref _freeStreamSlots);
```

Reply:
Good point. I revisited all concurrently accessed fields and hardened read/write operations with Interlocked or Volatile methods as appropriate.